anoma/namada-interface
# feat: proposals

## TODO:

- [x] check why quering crashes the tab
- [x] be sure that interface is connected to extension and addresses are present
- [x] move dialog to the separate file
- [x] replace Option with fp-ts option
- [x] rename Governance to Proposals
- [x] map rust query response type into proper type in shared?
- [x] update voting info after tx is finished
- [ ] ~~add loader?~~
- [ ] ~~add info about empty proposals~~
- [x] add toast
- [x] figure out why approve tx window does not close
- [ ] ~~create PR for changes in Namada~~

## Testing
- init and start chain as usual - it's good to  have longer epochs though
- create `proposal.json` file in the root of namada
```json
{
  "proposal": {
    "content": {
      "title": "TheTitle",
      "authors": "test@test.com",
      "discussions-to": "www.github.com/anoma/aip/1",
      "created": "2022-03-10T08:54:37Z",
      "license": "MIT",
      "abstract": "Ut convallis eleifend orci vel venenatis. Duis vulputate metus in lacus sollicitudin vestibulum. Suspendisse vel velit ac est consectetur feugiat nec ac urna. Ut faucibus ex nec dictum fermentum. Morbi aliquet purus at sollicitudin ultrices. Quisque viverra varius cursus. Praesent sed mauris gravida, pharetra turpis non, gravida eros. Nullam sed ex justo. Ut at placerat ipsum, sit amet rhoncus libero. Sed blandit non purus non suscipit. Phasellus sed quam nec augue bibendum bibendum ut vitae urna. Sed odio diam, ornare nec sapien eget, congue viverra enim.",
      "motivation": "Ut convallis eleifend orci vel venenatis. Duis vulputate metus in lacus sollicitudin vestibulum. Suspendisse vel velit ac est consectetur feugiat nec ac urna. Ut faucibus ex nec dictum fermentum. Morbi aliquet purus at sollicitudin ultrices.",
      "details": "Ut convallis eleifend orci vel venenatis. Duis vulputate metus in lacus sollicitudin vestibulum. Suspendisse vel velit ac est consectetur feugiat nec ac urna. Ut faucibus ex nec dictum fermentum. Morbi aliquet purus at sollicitudin ultrices. Quisque viverra varius cursus. Praesent sed mauris gravida, pharetra turpis non, gravida eros.",
      "requires": "2"
    },
    "author": "{ACCOUNT_ADDRESS_WITH_ENOUGH_TOKENS_TO_INIT_PROPOSAL}",
    "voting_start_epoch": 3,
    "voting_end_epoch": 18,
    "grace_epoch": 24
  }
}
```
Start epoch and end epoch has to be a  multiply of 3
Grace epoch has to be at least 6 more than end epoch
- `namada client init-proposal --data-path=proposal.json --signing-keys {your_account}`
- go to staking view and stake some tokens with validator - otherwise you won't have voting power
- ~~Not sure if this is intended but when you delegate tokens let's say in epoch 1, then your delegation is only visible on epoch 3, therefore your voting power is available during epoch 3 and not before - I will double check it~~ `there's "pipeline" offset of 2 epochs for most actions in PoS - the delegation should start contributing in epoch 3`
- go to proposals view, see initialized proposals list, click on proposal and scroll down to vote(if it is already ongoing)
